### PR TITLE
FileDescriptor: Add resize(to newSize:) and fileSize() APIs

### PR DESCRIPTION
**Pull request by simonjbeaumont**
The `FileDescriptor` API currently doesn't expose anything that resembles `ftruncate(2)` for truncating or extending the file associated with a file descriptor.

This PR adds a new API `FileDescriptor.resize(to newSize:)` which wraps `ftruncate`, available on non-Windows platforms.

### CODE REVIEW

File: Sources/System/Internals/WindowsSyscallAdapters.swift [hunk 1]
```diff
@@ -105,4 +105,11 @@ internal func pwrite(
   return Int(nNumberOfBytesWritten)
 }
 
+@inline(__always)
+internal func ftruncate(
+  _ fd: Int32, _ length: off_t
+) -> Int32 {
+  _chsize(fd, numericCast(length))
+}
```
Comment (simonjbeaumont) [inline Sources/System/Internals/WindowsSyscallAdapters.swift:113]: Full disclosure: I have not tested the Windows path. AFAICT there isn't a Windows CI pipeline?

**Comment by simonjbeaumont** [comment]
@swift-ci please test

**Comment by simonjbeaumont** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@lorentey can you take an initial look?

File: Sources/System/FileHelpers.swift [hunk 2]
```diff
@@ -122,4 +122,18 @@ extension FileDescriptor {
       return .success(buffer.count)
     }
   }
+
+  /// Return the current size of the file.
+  ///
+  /// - Returns: The current size of the file, in bytes.
+  @_alwaysEmitIntoClient
+  @discardableResult
```
Comment (simonjbeaumont) [inline Sources/System/FileHelpers.swift:130]: Looking at this again, this probably shouldn't be discardable.

File: Sources/System/FileHelpers.swift [hunk 3]
```diff
@@ -122,4 +122,18 @@ extension FileDescriptor {
...
+  public func fileSize(
+    retryOnInterrupt: Bool = true
+  ) throws -> Int64 {
+    let current = try seek(offset: 0, from: .current)
+    let size = try seek(offset: 0, from: .end)
+    try seek(offset: current, from: .start)
```
Comment (lorentey) [inline Sources/System/FileHelpers.swift:136]: This is temporarily mutating the file position during what looks like a read-only operation; I feel the potential pitfalls of this (alongside the fact that many file descriptors do not have a file size) probably makes this entry point less suitable for this package.

File: Sources/System/Internals/WindowsSyscallAdapters.swift (hunk 1 above)
Comment (lorentey) [reply Sources/System/Internals/WindowsSyscallAdapters.swift:113]: I believe it's best to surround the addition of `resize` with `#if !os(Windows)`.

Generally, I don't think it's a good idea to try to shoehorn Windows support into POSIX code paths here, or anywhere else in this code base. Swift System is not intended to provide a portable API layer, and the Windows SDK is different enough that it deserves its own, standalone, implementations.

**Review by lorentey — APPROVED** [review]
Big thumbs up on `resize`! I'm less convinced we need `fileSize`.

Let's keep the Windows implementation out for now.

File: Sources/System/FileHelpers.swift (hunk 3 above)
Comment (simonjbeaumont) [reply Sources/System/FileHelpers.swift]: That's reasonable. I added the following fixup to address this:

* d90bf85 `fixup: Move FileDescriptor.fileSize() to an internal extension in SystemTests`

File: Sources/System/Internals/WindowsSyscallAdapters.swift (hunk 1 above)
Comment (simonjbeaumont) [reply Sources/System/Internals/WindowsSyscallAdapters.swift]: OK, apologies for that, I was just following the pattern I saw with the existing `WindowsSyscallAdapters`. I added some fixups which move this API to only non-Windows platforms:

* 02daf4b `fixup: Move FileDescriptor.resize to extension with #if !os(Windows)`
* 7fab65f `fixup: Remove Windows syscall adaptor for ftruncate`

File: Sources/System/FileHelpers.swift (hunk 2 above)
Comment (simonjbeaumont) [reply Sources/System/FileHelpers.swift]: Addressed as part of:

* d90bf85 `fixup: Move FileDescriptor.fileSize() to an internal extension in SystemTests`

**Review by simonjbeaumont — COMMENTED** [review]
Thanks @lorentey for the review!

> Big thumbs up on `resize`! I'm less convinced we need `fileSize`.

No problem, I moved `fileSize` to be an implementation detail of the tests for `resize`.

> Let's keep the Windows implementation out for now.

Done.

File: Sources/System/FileOperations.swift [hunk 4]
```diff
@@ -398,3 +398,51 @@ extension FileDescriptor {
...
+  /// size of the file is smaller than `newSize`, the file is extended and filled with zeros to the
+  /// provided size.
+  ///
+  /// This function requires that the file has been opened for writing.
+  ///
+  /// - Note: This function does not modify the current offset for any open file descriptors
+  /// associated with the file.
+  ///
+  /// The corresponding C function is `ftruncate`.
+  /*System 1.2.0, @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)*/
```
Comment (milseman) [inline Sources/System/FileOperations.swift:425]: @lorentey what's the right availability comment here?

Comment (simonjbeaumont) [reply Sources/System/FileOperations.swift:425]: Yeah, I took a punt here based on what I saw with previous evolution and given there already exists a public `1.1.0` tag. LMK if it needs changing.

**Comment by simonjbeaumont** [comment]
@swift-ci please test

File: Sources/System/FileOperations.swift (hunk 4 above)
Comment (simonjbeaumont) [reply Sources/System/FileOperations.swift:425]: @lorentey Could you advise on what this magic comment string should be? I took a look at `./Utilities/expand-availability.py` and now I think maybe it should just be the following:

```swift
/*System 1.2.0*/
```

Happy to update this to reflect your release workflow if you can point me in the right direction here.

Comment (simonjbeaumont) [reply Sources/System/FileOperations.swift:425]: @lorentey LMK what this should be and I can update this PR.

Comment (lorentey) [reply Sources/System/FileOperations.swift:425]: Ah, apologies for the delay -- `/*System 1.2.0*/` is going to be the right one, it just needs to be defined in `/Utilities/expand-availability.py`. (Expanded to 9999 versions.)

We can define the version directly within this PR, or we can just land it with an undefined version, and let maintainers deal with it later. (These comments do not matter within this package variant of System.)

Comment (lorentey) [reply Sources/System/FileOperations.swift:425]: (To clarify, the comment we have here is just fine -- we don't need to remove the `@available` incantation.)

Comment (simonjbeaumont) [reply Sources/System/FileOperations.swift:425]: Added in:

* ca6e888 `Utilities: Add version 1.2.0 to expand-availability.py tool`

**Comment by simonjbeaumont** [comment]
@swift-ci please test

File: Sources/System/FileOperations.swift (hunk 4 above)
Comment (lorentey) [reply Sources/System/FileOperations.swift:425]: How did I miss it! 👍

**Comment by simonjbeaumont** [comment]
@swift-ci please test

**Pull request merged**